src/core/ddsi/src/q_xevent.c

```diff
@@ -594,55 +593,49 @@ static void send_heartbeat_to_all_readers(struct nn_xpack *xp, struct xevent *ev
   if (!writer_must_have_hb_scheduled (wr, &whcst))
   {
     hbansreq = 1; /* just for trace */
-    msg = NULL; /* Need not send it now, and no need to schedule it for the future */
     t_next.v = T_NEVER;
   }
   else if (!writer_hbcontrol_must_send (wr, &whcst, tnow))
   {
     hbansreq = 1; /* just for trace */
-    msg = NULL;
     t_next.v = tnow.v + writer_hbcontrol_intv (wr, &whcst, tnow);
   }
   else
   {
     struct wr_prd_match *m;
-    ddsrt_avl_iter_t it;
+    struct ddsi_guid last_guid = { .prefix = {.u = {0,0,0}}, .entityid = {0} };
 
     hbansreq = writer_hbcontrol_ack_required (wr, &whcst, tnow);
     t_next.v = tnow.v + writer_hbcontrol_intv (wr, &whcst, tnow);
 
-    for (m = ddsrt_avl_iter_first (&wr_readers_treedef, &wr->readers, &it); m; m = ddsrt_avl_iter_next (&it))
+    while ((m = ddsrt_avl_lookup_succ (&wr_readers_treedef, &wr->readers, &last_guid)) != NULL)
     {
```

Comment:
Methinks this is now an infinite loop: ``last_guid`` doesn't get updated. Adding ``last_guid = m->prd_guid;`` should solve that problem.